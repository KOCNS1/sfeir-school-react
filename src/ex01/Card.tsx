// define Card as a React component
// return the elements as in Card.html
// pass it a title prop

import React from "react";

type CardProps = {
  title: string;
};

export const test = "Toto";

export const Card: React.FC<CardProps> = (props) => {
  const styleCard = {
    width: 480,
  };

  return (
    <section className="mdc-card" style={styleCard}>
      <div style={{ padding: "1rem" }}>
        <h1 className="mdc-typography--headline4">{props.title}</h1>
      </div>
    </section>
  );
};

export default "Delfingen";
