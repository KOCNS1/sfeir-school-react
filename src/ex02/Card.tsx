import React from "react";

export const Card: React.FC = ({children}) => (
    <section className="mdc-card">
      <div className="card-content content-type-person-info">{children}</div>
    </section>
);

// implement these subcomponents

type CardImageProps = {
  url: string;
  desc: string;
};

export const CardImage: React.FC<CardImageProps> = ({url, desc}) => (
    <figure>
      <img src={url} alt={desc}/>
    </figure>
);

type CardHeaderProps = {
  title: React.ReactNode;
  subTitle: React.ReactNode;
};

export const CardHeader: React.FC<CardHeaderProps> = ({ title, subTitle}) => (
    <header>
      <h1 className="mdc-typography--headline5">
        {title}
      </h1>
      <h2 className="mdc-typography--subtitle1">{subTitle}</h2>
    </header>
);

interface CardInfoProps {
  icon: string;
  desc?: string;
}

export const CardInfo: React.FC<CardInfoProps> = ({ icon, children, desc }) => (
    <p>
      <i
          className="rmwc-icon material-icons rmwc-icon--size-small"
          title={desc}
      >
        {icon}
      </i>
      &nbsp;
      {children}
    </p>
);