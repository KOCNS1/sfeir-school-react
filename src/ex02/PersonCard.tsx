import React from "react";
import { Card, CardImage, CardHeader, CardInfo } from "./Card";

type PersonCardProps = {
  person: Person;
};

export const PersonCard: React.FC<PersonCardProps> = ({ person }) => (
  <Card>
    <CardImage url={person.photo} desc="face of Leanne" />
    <CardHeader
      title={
        <a href={`/person/${person.id}`}>
          {person.firstname} {person.lastname}
        </a>
      }
      subTitle={person.position}
    />
    <CardInfo icon="email">
      <a href={`mailto:${person.email}`}>{person.email}</a>
    </CardInfo>
    <CardInfo icon="phone">
      <a href={`tel:${person.phone}`}>{person.phone}</a>
    </CardInfo>
    {person.manager && (
      <CardInfo icon="supervisor_account" desc="manager">
        <a href={`/person/${person.managerId}`}>{person.manager}</a>
      </CardInfo>
    )}
  </Card>
);
