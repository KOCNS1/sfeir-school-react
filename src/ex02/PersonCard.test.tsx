import React from "react";
import renderer from "react-test-renderer";
import { PersonCard } from "./PersonCard";
import { Card } from "./Card";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import "regenerator-runtime/runtime";

// To set up Jest & tests

// 1) Installs deps
// > npm install -D react-test-renderer@16.13.0 @babel/preset-react @babel/preset-typescript

// 2) create .babelrc
// {
//     "presets": ["@babel/env", "@babel/react", "@babel/typescript"]
// }

// 3) Add script in package.json :
//   ...
//     "test": "jest --watchAll ./src"
//   ...

// 4) disable parcel babel integration ...
// go here -> https://parceljs.org/languages/javascript/#usage-with-other-tools

describe("PersonCard", () => {
  const person: Person = {
    id: "5763cd4d9d2a4f259b53c901",
    photo: "https://randomuser.me/portraits/women/85.jpg",
    firstname: "Leanne",
    lastname: "Woodard",
    position: "Developer",
    email: "woodard.l@acme.com",
    phone: "0784112248",
    manager: "Erika",
    managerId: "5763cd4d3b57c672861bfa1f",
  };

  const personWithoutManager: Person = {
    id: "5763cd4d9d2a4f259b53c901",
    photo: "https://randomuser.me/portraits/women/85.jpg",
    firstname: "Leanne",
    lastname: "Woodard",
    position: "Developer",
    email: "woodard.l@acme.com",
    phone: "0784112248",
    manager: "",
    managerId: "",
  };

  it("should render div", () => {
    const tree = renderer.create(<div>Hello !</div>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render Card", () => {
    const tree = renderer.create(<Card>hello card</Card>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render Person with manager", () => {
    const tree = renderer.create(<PersonCard person={person} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render Person without manager", () => {
    const tree = renderer
      .create(<PersonCard person={personWithoutManager} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // > npm install --save-dev regenerator-runtime

  // import { render, screen } from "@testing-library/react";
  // import "@testing-library/jest-dom";
  // import "regenerator-runtime/runtime";

  it("should render full name", async () => {
    render(<PersonCard person={person} />);

    const element = screen.getByTestId("full-name");

    expect(element).toHaveTextContent("Leanne Woodard");
  });

  // mail
  it("should render mail", async () => {
    render(<PersonCard person={person} />);

    const element = screen.getByTestId("mail");

    expect(element).toHaveTextContent(person.email);
  });

  // phone
  it("should render phone", async () => {
    render(<PersonCard person={person} />);

    const element = screen.getByTestId("phone");

    expect(element).toHaveTextContent(person.phone);
  });

  // manager
  it("should render manager", async () => {
    render(<PersonCard person={person} />);

    const element = screen.getByTestId("manager");

    expect(element).toHaveTextContent(person.manager);
  });

  it("should not render manager", async () => {
    render(<PersonCard person={personWithoutManager} />);

    const element = screen.queryByTitle("manager");

    expect(element).toBeNull();
  });
});
