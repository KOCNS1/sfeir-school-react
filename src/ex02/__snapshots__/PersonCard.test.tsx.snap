// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonCard should render div 1`] = `
<div>
  Hello !
</div>
`;
